### ISSUE TITLE
Add more flexibility for specification of object ID

### ISSUE THREAD

**Issue by rberlind** [Mastercard/terraform-provider-restapi#4]
When invoking Terraform Enterprise API endpoints, one typically gets back a response that looks like:

```
{
  "data": {
        "id": "4111797",
        "type": ...
        ...
   }
}
```
While I can create TFE objects with this provider, since the id is nested under data, the ID of the object is not set correctly in the Terraform state and it therefore cannot be destroyed.

Would it be possible to handle the ID deeper in the api_data structure?

Possibilities that occur to me include:
1. search for the id_attribute deeper in the structure returned by the API.
2. Let user specify full path to the ID, giving something like "data.id" or "employee.personal.ssn".
3. Convert id_attribute to an array and handle as many entries as user provides up to 3 or some other N.
4. Replace id_attribute with id_attribute_1, id_attribute_2, id_attribute_3.

Additionally, it would also be useful to allow the user to specify that the id_attribute is actually the last segment of the resource path.  This would be useful for APIs like Vault's API in which the object created such are often specified at the end of the URL rather than in the body of the request.  For more on this, see https://github.com/Mastercard/terraform-provider-restapi/issues/5

**Comment by rberlind** [comment]
It would also be useful to be able to obtain the ID from a specific header.  This is actually the most critical need for a customer I'm working with who is invoking the vRO REST API to run vRO workflows. https://docs.vmware.com/en/vRealize-Orchestrator/7.2/com.vmware.vrealize.orchestrator-develop-web-services.doc/GUID-5312E7E5-90B7-4D14-9BDA-43BF22E0DDF4.html indicates that POST requests to run new executions of workflows return the ID of the workflow instance in a link in the Location header of the response.  For instance, the Location header might be something like "https://myVROserver.xyz.com:8281/vco/api/workflows/06db9b35-bc27-4895-a878-942144d62790/executions/8ab79b96634d0d210163cfe7dd684401"

The actual ID of the created execution is the last segment of this URL, "8ab79b96634d0d210163cfe7dd684401".

So, the customer would need to be able to specify that the ID of the created execution was in the Location header and that the actual ID was the last segment of the URL.  The provider would then follow the post with a get against https://myVROserver.xyz.com:8281/vco/api/workflows/06db9b35-bc27-4895-a878-942144d62790/executions/8ab79b96634d0d210163cfe7dd684401.

If you did not want to bother trying to extract the ID from the Location header, perhaps you could add an attribute on the provider or resource that would give the URL to use for GET, PUT, and DELETE operations against a created resource after its initial creation and support getting that URL from a specific response header? After all, it is the URL that you ultimately need to use when invoking those methods.

Would implementing this in the next few days be possible?  Thanks so much.

Roger

**Comment by DRuggeri** [comment]
I will have to go back to the code and see if there's a clean way I can specify something deeper in the object structure. Part of the challenge is that we may have to invent a new syntax to find deep objects inside deep objects like `response.results|data.id` where the left side of the `|` is the location of the results array and the right side of the `|` is the location within each element. Doing something like that is my current thinking and it seems like a valid use case for a generic REST implementation.

Unfortunately, I can't think of a clean way to do the same from an arbitrary HTTP header. That seems a bit out of scope for this particular provider.

Availability to adjust and code something for this in the next few days is a bit rough. If #5 can get you over the hump, that would be much easier to write and release. Of course, PRs are always welcome if you would like to take a crack at it.

**Comment by rberlind** [comment]
Unfortunately, after thinking about the situation with respect to the vRO API, I realize that simply letting a user specify the get_path will not suffice.  The problem is that they won't know the right get_path before creating the resource.  For vRO, that get_path would have to be set somehow to the URL contained in the Location header returned by the POST to the create_path.

More generally, we might need to support setting read_path, update_path, and destroy_path to the value of some specific header or attribute of the JSON response returned by the POST. That would seem to require quite a bit of additional processing and probably additional attributes.

I apologize that this is all so horribly complicated.  I guess that either the "REST API" standard is not so standard after all, or that a lot of software vendors and developers are calling APIs "REST APIs" that are really just APIs that happen to send and return JSON and/or XML but might not conform to the standards.

Perhaps adding the additional URL attributes in #5 was a mistake aimed at giving a quick fix (which I very much appreciate)? Perhaps adding the additional flexibility of determining the object ID requested in this issue is the better approach.  If we make the specification of the object ID general enough and also support ID = "" in conjunction with certain other attributes being set, then the need for the extra URLs might go away.

**Comment by rberlind** [comment]
Sounds very good.  I'll check out more this weekend.

**Comment by DRuggeri** [comment]
I've added some test cases to cover this code as part of the datasource and using deeper `search_key` values, so I think it's safe to close this one out. Feel free to open it back up if you have additional input or if things don't quite work as you'd expect.

**Comment by rberlind** [comment]
Apologies for not following up on this @DRuggeri after you added the code over 3 weeks ago.  I got distracted and forgot all about this.  I appreciate your adding  support for id_attribute to be deeper in the structures and ended up seeing that when investigating the issue my PR for the data source was having. I even ended up using a deeper id_attribute "attributes/name" in my successful test of the data source after I made a modification to your 1.5 release.  So, I think we're good on this.  Thanks again.
